lib/psych/core_ext.rb

```diff
@@ -18,12 +18,19 @@ def to_yaml options = {}
   require_relative 'y'
 end
 
-
-# TODO: how best to check for builtin Set?
-if defined?(::Set) && Object.const_source_location(:Set) == ["ruby", 0]
+# Up to Ruby 3.4, Set was a regular object and was dumped as such
+# by Pysch.
+# Starting from Ruby 3.5 it's a C extension, so we have to implement
+# #encode_with / #init_with to preserve backward compatibility.
+if defined?(::Set) && !Psych.dump(Set.new([42])).include?("42")
```

Comment:
```suggestion
if defined?(::Set) && !Psych.dump(Set[42]).include?("42")
```

Just a little simpler.

Reply:
Why not check for @hash ivar?

Reply:
I was worried it might not behave correctly on other implementations, but yes, it's probably the simpler